crates/telio-starcast/src/transport.rs

```diff
@@ -0,0 +1,231 @@
+    /// * socket_pool - To create the transport socket
+    /// * packet_chan - A channel to send packets to and receive packets from the transport component
+    /// multicast_ips - IPs that native multicast packets can be sent to, e.g. 224.0.0.0/4
+    pub async fn start(
+        vpeer_ip: DualIpAddr,
+        meshnet_ip: IpAddr,
+        socket_pool: Arc<SocketPool>,
+        packet_chan: Chan<Vec<u8>>,
+        multicast_ips: Vec<IpNetwork>,
+    ) -> Result<Self, Error> {
+        let transport_socket = socket_pool
+            .new_internal_udp(SocketAddr::new(meshnet_ip, MULTICAST_TRANSPORT_PORT), None)
+            .await
+            .map_err(|e| {
+                telio_log_error!("Failed to bind multicast socket: {:?}", e);
+                Error::SocketBindError(e.to_string())
+            })?;
+        let transport_socket = Arc::new(transport_socket);
+
+        Ok(Self {
+            task: Task::start(State {
+                transport_socket,
+                packet_chan,
+                multicast_ips,
+                recv_buffer: vec![0; MAX_PACKET_SIZE],
+                nat: StarcastNat::new(vpeer_ip.ipv4, vpeer_ip.ipv6),
+                peers: Vec::new(),
+            }),
+        })
+    }
+
+    /// Update runtime values of the transport component
+    pub async fn configure(&self, config: Config) -> Result<(), Error> {
+        task_exec!(&self.task, async move |state| {
+            state.configure(config);
+            Ok(())
+        })
+        .await?;
+        Ok(())
+    }
+
+    /// Stop the transport component
+    pub async fn stop(self) {
+        let _ = self.task.stop().await.resume_unwind();
+    }
+}
+
+struct State {
+    transport_socket: Arc<UdpSocket>,
+    packet_chan: Chan<Vec<u8>>,
+    recv_buffer: Vec<u8>,
+    nat: StarcastNat,
+    peers: Vec<Peer>,
+    multicast_ips: Vec<IpNetwork>,
+}
+
+impl State {
+    fn configure(&mut self, config: Config) {
+        self.peers = config.peers;
+    }
+
+    async fn handle_local_multicast_packet(&mut self, packet: Vec<u8>) -> Result<(), Error> {
+        let failed_peers = join_all(self.peers.iter().map(|peer| {
+            self.transport_socket
+                .send_to(
+                    &packet,
+                    SocketAddr::new(peer.addr, MULTICAST_TRANSPORT_PORT),
+                )
+                .map_err(|_| peer.public_key)
+        }))
+        .await
+        .into_iter()
+        .filter_map(|res| match res {
+            Ok(_) => None,
+            Err(pk) => Some(pk),
+        })
+        .collect::<Vec<_>>();
+
+        if failed_peers.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::FanoutFailed(failed_peers))
+        }
+    }
+
+    async fn handle_mapped_unicast_packet(&mut self, mut packet: Vec<u8>) -> Result<(), Error> {
+        self.nat
+            .translate_outgoing(&mut packet)
+            .map_err(Error::NatError)?;
+        self.transport_socket
+            .send(&packet)
```

Comment:
Where are you sending this packet?

transport socket is not connected to anything this will fail.
you need to use send to, and get dst ip as discussed

Reply:
I also forgot to return this address from translate ongoing. I wanted to do this during the integration. Those are the changes into `nat.rs`:
```
diff --git a/crates/telio-starcast/src/nat.rs b/crates/telio-starcast/src/nat.rs
index 9d5dc868c..21c0322d5 100644
--- a/crates/telio-starcast/src/nat.rs
+++ b/crates/telio-starcast/src/nat.rs
@@ -30,7 +30,7 @@ pub trait Nat {
     fn translate_incoming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
     /// Translate outgoing packet (from the multicast peer to the transport socket)
     /// Change the destination to the peer's original ip and port
-    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<IpAddr, Error>;
 }
 
 type Mapper = fn(src_addr: SocketAddr) -> u16;
@@ -140,7 +140,7 @@ impl StarcastNat {
     fn translate_outgoing_internal<'a, P: MutableIpPacket<'a>>(
         &mut self,
         packet: &'a mut [u8],
-    ) -> Result<(), Error> {
+    ) -> Result<IpAddr, Error> {
         let mut ip_packet = P::new(packet).ok_or(Error::PacketTooShort)?;
         if ip_packet.get_next_level_protocol() != IpNextHeaderProtocols::Udp {
             return Err(Error::UnexpectedTransportProtocol);
@@ -171,7 +171,7 @@ impl StarcastNat {
         ip_packet.set_destination(new_dst_ip);
         ip_packet.fix_ip_header_checksum(old_dst_ip, new_dst_ip);
 
-        Ok(())
+        Ok(new_dst_ip.into())
     }
 }
 
@@ -187,7 +187,7 @@ impl Nat for StarcastNat {
         }
     }
 
-    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error> {
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<IpAddr, Error> {
         match packet.first().ok_or(Error::PacketTooShort)? >> 4 {
             4 => self.translate_outgoing_internal::<MutableIpv4Packet>(packet),
             6 => self.translate_outgoing_internal::<MutableIpv6Packet>(packet),
```
As @packgron said you need to take this IP address and combine it with transport port to get SocketAddr to use in `sendto`